timemachine/fe/rbfe.py

```diff
@@ -199,41 +213,73 @@ def setup_initial_states(
     # perform any concluding sanity-checks
     assert_all_states_have_same_masses(initial_states)
 
-    return initial_states
+    def make_optimized_initial_state(lamb: float) -> InitialState:
+        initial_state = make_initial_state(lamb)
+
+        # use pre-optimized initial state with the closest value of lambda as a starting point for optimization
+        nearest_optimized = min(initial_states, key=lambda s: abs(lamb - s.lamb))
```

Comment:
Looking further at the code it looks like since there is a grid of lambda generated for this in the hot path this is not a big deal. It does makes this code confusing as implicitly you should construct this with a grid to get what I would expect for the intermediate poses. Having a comment that suggests generating with more then the endstates might be a good idea.

Reply:
Right, the selection of the nearest initial state here is from the grid of `n_windows` states we optimize at the outset (rather than from the states generated by bisection). This is a bit confusing, and maybe slightly suboptimal in that we don't consider optimized conformers generated by bisection (which might be the closest state available after many bisections)

Reply:
Having a warning in the docstring for the function would be fine. 

Maybe something like the following would work as well, though its hacky
```
def scoring_func(s):
    diff = lamb - s.lamb
    if lamb < 0.5:
        # If lamb is <0.5, prefer frames from the the 0.0 endstate
        # else prefer larger lambda states from the 1.0 endstate
        diff = abs(diff)
    return diff

Reply:
Hmm, I'm not sure I understand; I think the situation you describe with a new lambda at 0.5 and only [0, 1] pre-optimized is not possible with the current logic. We'd only consider adding a 0.5 state if `n_windows >= 3`, in which case we'd have at least one pre-optimized intermediate. Am I missing the point?

Reply:
The issue is that this function has subtly different behavior depending on how you use it. The way you use in the PR is correct, but it is not obvious that you need to use it that way. If you do the following, the poses will be biased towards the 0.0 endstate (if across python versions min() returns the first lowest instance, which seems likely).

```
states, make_state = setup_initial_states(..., lambda_schedule=[0.0, 1.0])
while lowest_ddg_err_of_states(states) < 0.618:
        new_good_lambda = my_special_func(states)
        states.append(make_state(new_good_lambda))
```

This was originally the way I thought I could use this function and it seemed like it would eventually be desirable in its usage. It seems like the function should be agnostic to its set up when it comes to generating poses as there is no restriction in the number of times you can create new states.

Reply:
Ah, I see, thanks for clarifying. I hadn't intended this kind of usage, i.e. mutating `initial_states` outside of the function and calling `make_state` again with it using the mutated version -- sometimes I forget that Python lets you do stuff like that! This behavior feels kind of dangerous to me even if it might be useful in some cases, so I'd probably prefer to make it impossible, e.g. by returning a copy of `initial_states` instead of the one in the closure of `make_optimized_initial_state`. Does that sound OK to you?

Reply:
Also not quite clear on how to think about the following, might not matter at all since you preserve the poses from the endstates (if you provide a grid of at least 3). Just know we currently do it iteratively and in order so I put some weight on that choice.

```
'->' means it was minimized from lower lambda to higher lambda value
'<-' means it was minimized from higher lambda to lower lamdba value

What we have currently
[0.0, ->, ->, ->, <-, <-, <-, <-, 1.0]
                 ^
               0.5 lamb

With this PR:
[0.0, ->, ->, ->, <-, ->, <-, ->, 1.0]
                 ^
               0.5 lamb
```
We are now minimizing in the reverse direction for lambda > 0.5. And I don't think the minimization is reversible.

Reply:
Changed to return a copy of `initial_states` to prevent unintended usage like the example above in https://github.com/proteneer/timemachine/pull/959/commits/fc47c1271598f9caa4dec4ca0a007fc70d89590e.

Question: is there any reason to prefer a starting conformation other than that with the closest lambda value for the fine-tuning optimizations done during bisection? (E.g. should we prefer use a starting conformation closer to an end-state?)

Reply:
So if you have an initial schedule of [0, 0.49, 1.0], then go to add 0.55, it will take the point from 0.49 and not 1.0, which is different from the current behavior.

@badisa I think this is what you are concerned about?

As n_windows increases, this becomes less of an issue, since we'd already have multiple lambdas near 0.5. In practice, I doubt it's going to be an issue if n_windows is large enough, but we could change it to make it consistent.

Reply:
That is what I was concerned about.

There is also the question of whether minimizing along `1.0 -> 0.8 -> 0.9` (what this PR will do) is different from minimizing along `1.0 -> 0.9 -> 0.8` (previous behavior). As any window that isn't in the pre-existing grid, will be minimized from the lower lambda value regardless of lambda value.

Reply:
> As any window that isn't in the pre-existing grid, will be minimized from the lower lambda value regardless of lambda value.

Note, this is only true if the proposed lambda is exactly between 2 lambdas in the initial linspace. In general (unless n_windows is a power of 2) the nearest lambda should be larger ~half of the time

Reply:
Right, keep forgetting that.

This makes me think we separate out this function for adding a new state into a separate function where you provide a set of states. That way the states you are using to generate a new state is explicit rather then implicit and you can more generically use this function without having to be aware of the internal state of the function.

This is a nit, but I have a hard time wrapping my head around this closure while reading this PR, and I imagine I will struggle with it later.

Reply:
It seems like we should fix this, since it seems simple. And it feels like a code smell to have the protocol behavior change based on the number of windows. While 2^k + 1 is not a window value I would typically expect, would rather not have to keep this detail in the back of my mind when running jobs

Reply:
@badisa to be clear, your proposed fix is, in the case where we have 2^k + 1 windows so there's an ambiguity in whether a bisection proposal should choose a pre-optimized conformer from the left or right, to go left when $\lambda < 0.5$ and right when $\lambda >= 0.5$?

Reply:
Exactly

Reply:
Implemented in https://github.com/proteneer/timemachine/pull/959/commits/52fd65637d766f792ce1a90f1d6d066c36f7fa66; @badisa would you mind taking a look?

Reply:
Looks great!

Reply:
Note: I realized I think there's some ambiguity in the above re whether we want to _always_ prefer the left of the 2 closest states for lambda <= 0.5, versus adding this bias only when the 2 closest states are equidistant. The PR currently does the latter. Thoughts @badisa @jkausrelay ?

Reply:
I would be inclined to *always* prefer the left or right based on the lambda value, to be consistent with the current implementation. 

This still behave differently in the case @jkausrelay had described https://github.com/proteneer/timemachine/pull/959#discussion_r1121929540 though that is not the typical place. Though fixing it now means we don't have to worry about it later if we want to start from a more optimized schedule.